url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/PoulidorDistanceFraudSimulator.java

```diff
@@ -0,0 +1,189 @@
+package protocols.specifications;
+
+import java.security.SecureRandom;
+import java.util.Hashtable;
+import java.util.TreeMap;
+
+
+/*Trujillo- Mar 11, 2014
+ * This simulator is similar to an exhaustive search. However, it limits
+ * the number of sequences kept in memory. Instead of having all of them, 
+ * we just keep the ones that are repeated more often.*/
+public class PoulidorDistanceFraudSimulator{
+	
+	public static final int RUNS = 1000000;
+	public static final int DEFAULT_SIZE = 65536;
+	
+	private int size = DEFAULT_SIZE;
+
+			
+	public PoulidorDistanceFraudSimulator(){			
+	}
+	
+	public double computeDistanceFraud(int n){
+		double result = 0;
+		for (int i = 0; i < RUNS; i++) {
+			SequenceInGraph mostFrequentSequence = runForMostFrequentSequence(n);
+			//System.out.println(mostFrequentSequence.toString());
+			result += (double)mostFrequentSequence.getTotalFrequency()/Math.pow(2, n);
+		}
+		return result/RUNS;
+	}
+	
+	private SequenceInGraph runForMostFrequentSequence(int n){
+		TreeMap<SequenceInGraph, SequenceInGraph> sequences = new TreeMap<>();
+		SequenceInGraph first = new SequenceInGraph("", -1, 1);
+		sequences.put(first, first);
+		boolean[] graph = generateRandomGraph(n);
+		//boolean[] graph = new boolean[]{false, true, false, false};
+		//System.out.println("Graph: "+ToStringUtility.toStringBooleanArray(graph));
+		sequences = runForMostFrequentSequence(n, sequences, graph);
+		//System.out.println("Frequency: "+sequences.firstKey().getTotalFrequency());
+		return sequences.firstKey();
+	}
+	
+	/*Trujillo- Mar 11, 2014
+	 * In this method, we have a list of sequences and their number of appearances. 
+	 * Different to the tree, in this method we keep the positions (the nodes) in the tree of
+	 * the sequences. Together with the position, we put the number of walks that generate 
+	 * this sequence until this position. 
+	 * So, let "s" a sequence and p1, ..., pm and f1 ... fm the positions and sequences
+	 * respectively for s. Each position pi generates two new positions pi+1 and pi+2 and
+	 * also two new (potentially equal) sequences. The frequency of both sequences is fi
+	 * regardless whether they are equal or not.
+	 * Frequencies are merged if a sequence "s" at some position "p" already exists
+	 * in the table. In this case both frequencies are summed. 
+	 * 
+	 * We always keep the table with a size lower than SIZE.
+	 * */
+	
+	/*Trujillo- Mar 11, 2014
+	 * TODO: Change the tree map by a sorted list using the class
+	 * SequenceInTree.*/
+	private TreeMap<SequenceInGraph, SequenceInGraph> runForMostFrequentSequence(int n, 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.